specs/Standards/MultiToken/README.md

```diff
@@ -0,0 +1,539 @@
+The rationale for this design, is that we want to support developers that need batch requests. In many ecosystems, a single NFT or 
+series of NFT are not sufficient for representing and managing potentially 100s of tokens. Managing cross contract calls and the token contracts themselves become a huge burden on the developer. This will reduce the complexity, and allow developers to easily bridge over to other contracts like ERC-1155 on other chains that allow for a this style of representation.
+
+In the design phase it was considered to simply rely on the underlying implementations of fungible tokens and non-fungible tokens
+to be the scope of interaction with the chain. Doing this we would have tied the implementations to the bounds of FT and NFT. By loosing this up a bit, we are able to be a bit more flexible in what's possible.
+
+Not doing this means we really won't have a great way of supporting use cases where developers need to represent and manage
+large varying quantities and types of tokens. They will all have to implement it on their own, and that would make for a fragmented
+and weak ecosystem. Where every developer would not be able to reliably trade these assets.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+The unresolved questions, are really what type of metadata , is required for this spec?
+- We decided that having metadata on chain is good, and that at later time we can have a spec for extra data if needed
+
+Can we represent events in this spec, which would improve the ecosystem quite a bit? If we represent events what should those events be?
+- We decided events aren't here yet in the ecosystem, in the way we' like them to be so they are't apart of this standard
+
+Should we have a spec for TokenType?
+- We decided TokenType shouldn't be exposed for the public interface consumption. We will leave this description in the hands of the implementers
+
+Should we have a spec for offchain metadata?
+- We decided no spec for offchain metadata yet but maybe in the future
+
+Does the current storage management scheme work for people?
+- The current storage management scheme works for folks
+
+How freeform should this token be? Right now there is a notion of supply, which is not 100% guaranteed every token has
+or wants to track supply semantics. Not having supply makes everything more difficult and requires consumers of the contract
+to track minting and burning events.
+- This point was taken to show that due to lack of proper events, and having the additional capability of storing this on chain, would
+result in a reduction of complexity for consumers of the contract data.
+
+Approval Management is probably out of the scope of this solution.
+
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+Future possibilities could be around enumeration extension like for the NFT spec, an off chain data spec, approval management if it's required, and error status codes.
```

Comment:
If we don't define enumeration, this spec isn't really usable from the wallet's perspective; without enumeration we can't show you tokens you own.  Am I missing something here?  Also, without enumeration implemented, eventually owners of large numbers of tokens will hit the viewFunction gas limit just trying to load their tokens list... IMO enumeration in a paginated way is critical to the viability of the standard.

Reply:
Hey good points here, for enumeration what would you be expecting to have access to that's not included within the spec. 

We assumed that the wallet team only needed to be able to be able to index mt/nft/ft prefixed items in order to be ready.

I'm linking a doc here https://hackmd.io/tTHosvAHTjeEHRTFR7Vj1A for ease that puts together some thoughts. I'm also inlining the same thoughts.

Here's some hot takes on this , this might be wallet slanted

#### `nft_total_supply(&self) -> U128`
```
/// Returns the total supply of non-fungible tokens as a string representing an
/// unsigned 128-bit integer to avoid JSON number limit of 2^53.
    fn nft_total_supply(&self) -> U128;  
```    
This is a bit strange from the multi token standard perspective, 
if you have a token_id then you can retrieve the total supply for that token_id, but you could have 1000s of different types of tokens within the same contract , additionally there's no coding in a grouping factor. 

So for instance there's no sense of owning 1000 bored apes. That series grouping is handled implicitly by metadata of the token, so consuming apps would have to decide how they want to group things. I think within the NFT enumeration there's some implicit grouping caused by being associated with the contract that doesn't hold for multi token standard.

I think the second thing here is to see that token types have different concepts of supply, it's not necessarily as straight forward. So a video game concept might 
have a few fungible coin tokens with supply 1million, they might have a few 
non fungible tokens like 10000 unique golden swords with serial numbers,
and a few gemstones 200 supply. 

How should we represent these? In theory we could have  1 token_id 1 count
Total supply could be  1,100,200 or 100,201 or some grouping of some sort

or 1 token_id 1 count, unelss it has a greator supply , so it 

we do have
```
 mt_total_supply(&self, token_id) -> U128
``` 
 but the behavior may vary if there is an assumption that nft_total_supply
 talks about one group of token_ids
 ``` 
 alternatively:
 mt_token_groups(&self, group_id: String, from_index:Option<U128>, limit: Option<U64>) -> Vec<String> 
 mt_total_supply_group(&self, group_id) -> U128
 ```
 
 This would allow consumers that want to group related token ids to be able to do that as apart of enumeration.  I think the way that devs avoid multiple storage cost is to simply return token_id as group_id, and for the interface mt_total_supply_group, just call  `mt_total_supply(&self, token_id)` .
 
 #### `fn nft_tokens(&self, from_index, limit) -> Vec<Token>` 

 ``` 
 /// Get a list of all tokens

///

/// Arguments:

/// * `from_index`: a string representing an unsigned 128-bit integer,

/// representing the starting index of tokens to return

/// * `limit`: the maximum number of tokens to return

///

/// Returns an array of Token objects, as described in Core standard

fn nft_tokens(

&self,

from_index: Option<U128>, // default: "0"

limit: Option<u64>, // default: unlimited (could fail due to gas limit)

) -> Vec<Token>;
 ```
 nft_tokens can be represented as mt_tokens(&self, from_index, limit) -> Vec<"Token">
 this is maybe a little different. is there any particular aim or usecase for this? An alternative direction would be to have a group by group grouping of this.
 
 #### `fn nft_supply_for_owner(&self, account_id: AccountId) -> U128;`
 ```
/// Get number of tokens owned by a given account
///
/// Arguments:
/// * `account_id`: a valid NEAR account
///
/// Returns the number of non-fungible tokens owned by given `account_id` as
/// a string representing the value as an unsigned 128-bit integer to avoid JSON
/// number limit of 2^53.   

fn nft_supply_for_owner(&self, account_id: AccountId) -> U128;
```
this method might be different, as there can be many different tokens would the intent here be to retrieve supply of for a particular token, as there might be fungible tokens
mixed in. We currently have 
```
mt_balance_of(&self,account_id,Vec<TokenId>)
```
There is no explict owner has these token ids. perhaps the alt or equivalent from
consuming perspective might be
```
struct TokenSupply {
tokenId: TokenId
supply : U128
}
// You may run out of gas for this so
mt_supply_for_owner(&self, account_id: AccountId, from_index: U128, limit: U64)->Vec<TokenSupply>
```
#### `nft_tokens_for_owner(self, account_id, from_index, limit)->Vec<Token>`
```
/// Get list of all tokens owned by a given account

/// Arguments:

/// * `account_id`: a valid NEAR account

/// * `from_index`: a string representing an unsigned 128-bit integer,

/// representing the starting index of tokens to return

/// * `limit`: the maximum number of tokens to return

///

/// Returns a paginated list of all tokens owned by this account

fn nft_tokens_for_owner(

&self,

account_id: AccountId,

from_index: Option<U128>, // default: "0"

limit: Option<u64>, // default: unlimited (could fail due to gas limit)

) -> Vec<Token>;

}
```
So for this we can do the same thing we did for mt_supply_for_owner only returning back 
Vec`<Token>` this has the same limits as well, where  you might intend to organize these together.
```
mt_tokens_for_owner(&self, account_id:AccountId, from_index, limit) -> Vec<Token>
```



@MaximusHaximus  what about the enumeration standard for nft's would you like to see does the above cover your concerns or needs? Are there other consumers like indexer and ecosystem participants that this helps. Finally is this the minimal set of things needed? 

We're headed towards another meeting to just get clarity. 
https://github.com/shipsgold/multi-token-standard-impl/blob/main/meetings/2021-09-13-meeting-agenda.md

Reply:
The other thing here for reference is that we did decide a while back to punt on enumeration for another standard post this, at future point in time. 

However, it seems like you see this as being problematic, as it will immediately prevent wallet team from doing anything with this spec. The above is mostly a few thoughts on the existing enumeration spec, and what it might mean for the token.

Reply:
Imagine for a moment that there is a smart contract wherein Bob owns some quantities of 5 different kinds of FTs and some quantities of 5 different NFTs.

Now, open the wallet and ask yourself "What RPC calls should the wallet dispatch in order to identify in an efficient way that this account owns 5 NFTs and some quantities of 5 FTs from a particular MT contract, then display them in the correct parts of the UI for the user?"  Right now there is nothing in the standard at all that answers this question; as far as I can tell, the only way anyone can tell what they own is by looking through transaction histories?

If a user can't query the contract to identify _which token Ids a user owns_ or _what kinds of tokens those token IDs correspond to_, how can this standard be _practically_ be used by anyone?  If everyone who _uses_ this standard has to roll their own query-ability, then all viewing of data about MT will be fragmented and inconsistent.  IMO there is very little value of having a standard that only defines transferring tokens and doesn't address how to query those tokens.

Given the current interface, it appears that we need to know specific token IDs to do _any querying of balances_ from the smart contract.

The key questions to answer are:
1. How does someone identify which token IDs that a particular account owns some quantity of?
2. How does someone know which of those tokens are NFTs?
3. How does someone know which of those tokens are FTs?
4. How does someone load the details of the tokens from #2 and #3 in a scalable and consistent way?

Without these questions having answers in the standard, the standard will not be usable and people will write their own view layer leading to inconsistencies across implementations of the standard.

Displaying and working with FTs is a _fundamentally_ different UI flow than NFTs due to their fungibility; if we can't quickly identify FTs vs. NFTs and paginate through them, how can the wallet show the user what they own?  If a wallet can't do it using standard RPC methods, then the _user_ can't do it themselves either, even using the CLI.

While the standard does a good job of describing how one would interact with the multi token smart contract to _transfer_ tokens, without Enumeration being defined, we are missing any guidance on how someone (or a wallet app) would interact with the same smart contract in order to _view_ what a particular user owns, either via the CLI or with an intent display it to them in a sane way in a UI like the wallet.

Reply:
Without proper enumeration, I think the absolute bare minimum would be a way to get a list of token IDs that a particular account owns, and whether each token ID is a FT or NFT type token.

However, as I mentioned, loading all tokens for a user in a single call without pagination will eventually break due to gas limits on view calls, so I would consider some kind of basic enumeration with pagination as a requirement for the standard if it was me.

Reply:
I think this has been addressed Events, Metadata, and Enumeration standards that have been added. PTAL